lib/src/json_schema/json_schema.dart

```diff
@@ -630,9 +630,8 @@ class JsonSchema {
             }
           }
 
-          // If currentSchema is a ref, resolve ref recursively.
-          // Cycle detection happens at evaluation time for drafts 2019 and 2020.
-          if (currentSchema.schemaVersion < SchemaVersion.draft2019_09 && currentSchema.ref != null) {
+          // If currentSchema is a ref, we might want to resolve it recursively right now.
+          if (currentSchema.ref != null && currentSchema._schemaMap.length == 1) {
```

Comment:
Why do we care whether the current schema has only one item? Could it not contain a ref, and also some keys (2019 and later)?